File: ChartsDemo/Classes/RealmDemos/RealmWikiExampleChartViewController.m
//
//  RealmWikiExampleChartViewController.m
//  ChartsDemo
//
//  Created by Daniel Cohen Gindi on 17/01/2016.
//
//  Copyright 2015 Daniel Cohen Gindi & Philipp Jahoda
//  A port of MPAndroidChart for iOS
//  Licensed under Apache License 2.0
//
//  https://github.com/danielgindi/Charts
//

#import "RealmWikiExampleChartViewController.h"
#import "ChartsDemo-Swift.h"
#import <Realm/Realm.h>
#import "Score.h"

@interface RealmWikiExampleChartViewController () <ChartViewDelegate>

@property (nonatomic, strong) IBOutlet LineChartView *lineChartView;
@property (nonatomic, strong) IBOutlet BarChartView *barChartView;

@end

@implementation RealmWikiExampleChartViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"Realm.io Wiki Example";
    
    _lineChartView.delegate = self;
    _barChartView.delegate = self;
    
    [self setupBarLineChartView:_lineChartView];
    [self setupBarLineChartView:_barChartView];
    
    _lineChartView.extraBottomOffset = 5.f;
    _lineChartView.extraRightOffset = 15.f;
    _barChartView.extraBottomOffset = 5.f;
    _barChartView.extraRightOffset = 15.f;
    
    _lineChartView.leftAxis.drawGridLinesEnabled = NO;
    _lineChartView.xAxis.drawGridLinesEnabled = NO;
    _barChartView.leftAxis.drawGridLinesEnabled = NO;
    _barChartView.xAxis.drawGridLinesEnabled = NO;

    // setup realm
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm beginWriteTransaction];
    
    // clear previous scores that might exist from previous viewing of this VC
    [realm deleteObjects:Score.allObjects];
    
    // write some demo-data into the realm.io database
    Score *score1 = [[Score alloc] initWithTotalScore:100.f scoreNr:0 playerName:@"Peter"];
    [realm addObject:score1];
    
    Score *score2 = [[Score alloc] initWithTotalScore:110.f scoreNr:1 playerName:@"Lisa"];
    [realm addObject:score2];
    
    Score *score3 = [[Score alloc] initWithTotalScore:130.f scoreNr:2 playerName:@"Dennis"];
    [realm addObject:score3];
    
    Score *score4 = [[Score alloc] initWithTotalScore:70.f scoreNr:3 playerName:@"Luke"];
    [realm addObject:score4];
    
    Score *score5 = [[Score alloc] initWithTotalScore:80.f scoreNr:4 playerName:@"Sarah"];
    [realm addObject:score5];
    
    // commit changes to realm db
    [realm commitWriteTransaction];
    
    // add data to the chart
    [self setData];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setData
{
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    // Line chart
    RLMResults *results = [Score allObjectsInRealm:realm];
    
    RealmLineDataSet *lineDataSet = [[RealmLineDataSet alloc] initWithResults:results yValueField:@"totalScore" xIndexField:@"scoreNr"];
    lineDataSet.drawCubicEnabled = NO;
    lineDataSet.label = @"Realm LineDataSet";
    lineDataSet.drawCircleHoleEnabled = NO;
    [lineDataSet setColor:[ChartColorTemplates colorFromString:@"#FF5722"]];
    [lineDataSet setCircleColor:[ChartColorTemplates colorFromString:@"#FF5722"]];
    lineDataSet.lineWidth = 1.8f;
    lineDataSet.circleRadius = 3.6f;
    
    NSArray<id <IChartDataSet>> *lineDataSets = @[lineDataSet];
    
    RealmLineData *lineData = [[RealmLineData alloc] initWithResults:results xValueField:@"playerName" dataSets:lineDataSets];
    [self styleData:lineData];
    
    // set data
    _lineChartView.data = lineData;
    [_lineChartView animateWithYAxisDuration:1.4
                                easingOption:ChartEasingOptionEaseInOutQuart];
    
    // Bar chart
    RealmBarDataSet *barDataSet = [[RealmBarDataSet alloc] initWithResults:results yValueField:@"totalScore" xIndexField:@"scoreNr"];
    barDataSet.colors = @[
                          [ChartColorTemplates colorFromString:@"#FF5722"],
                          [ChartColorTemplates colorFromString:@"#03A9F4"],
                          ];
    barDataSet.label = @"Realm BarDataSet";
    
    NSArray<id <IChartDataSet>> *barDataSets = @[barDataSet];
    
    RealmBarData *barData = [[RealmBarData alloc] initWithResults:results xValueField:@"playerName" dataSets:barDataSets];
    [self styleData:barData];
    
    _barChartView.data = barData;
    [_barChartView animateWithYAxisDuration:1.4
                                easingOption:ChartEasingOptionEaseInOutQuart];
}

#pragma mark - ChartViewDelegate

- (void)chartValueSelected:(ChartViewBase * __nonnull)chartView entry:(ChartDataEntry * __nonnull)entry dataSetIndex:(NSInteger)dataSetIndex highlight:(ChartHighlight * __nonnull)highlight
{
    if (chartView == _lineChartView)
    {
        NSLog(@"chartValueSelected in Line Chart");
    }
    else
    {
        NSLog(@"chartValueSelected in Bar Chart");
    }
}

- (void)chartValueNothingSelected:(ChartViewBase * __nonnull)chartView
{
    if (chartView == _lineChartView)
    {
        NSLog(@"chartValueNothingSelected in Line Chart");
    }
    else
    {
        NSLog(@"chartValueNothingSelected in Bar Chart");
    }
}

@end

File: ChartsDemo-OSX/ChartsDemo-OSX/RealmDemo/RealmWikiExampleChartViewController.swift
//
//  RealmBarChartViewController.swift
//  ChartsDemo-OSX
//
//  Created by thierryH24A on 02/04/2017.
//  Copyright © 2017 dcg. All rights reserved.
//

import Cocoa
import Charts
import Realm
import ChartsRealm
import RealmSwift

class RealmWikiExampleChartViewController: RealmDemoBaseViewController {
    
    @IBOutlet var lineChartView: LineChartView!
    @IBOutlet var barChartView: BarChartView!
    
    var results: RLMResults<Score>!
    weak var axisFormatDelegate: IAxisValueFormatter?
    
    override open func viewDidAppear()
    {
        super.viewDidAppear()
        view.window!.title = "Realm.io Wiki Example"
    }
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        writeRandomDataToDb(withObjectCount: 40)
        title = "Realm.io Wiki Example"
        
        _ = [["key": "toggleValues", "label": "Toggle Values"],
             ["key": "toggleHighlight", "label": "Toggle Highlight"],
             ["key": "animateX", "label": "Animate X"],
             ["key": "animateY", "label": "Animate Y"],
             ["key": "animateXY", "label": "Animate XY"],
             ["key": "saveToGallery", "label": "Save to Camera Roll"],
             ["key": "togglePinchZoom", "label": "Toggle PinchZoom"],
             ["key": "toggleAutoScaleMinMax", "label": "Toggle auto scale min/max"]]
        
        axisFormatDelegate = self
        
        setupBarLineChartView(lineChartView)
        lineChartView.delegate = self
        lineChartView.extraBottomOffset = 5.0
        lineChartView.leftAxis.drawGridLinesEnabled = false
        lineChartView.xAxis.drawGridLinesEnabled = false
        lineChartView.xAxis.labelCount = 5
        lineChartView.xAxis.granularity = 1.0
        lineChartView.xAxis.labelPosition = .bottom
        
        setupBarLineChartView(barChartView)
        barChartView.delegate = self
        barChartView.extraBottomOffset = 5.0
        barChartView.leftAxis.drawGridLinesEnabled = false
        barChartView.xAxis.drawGridLinesEnabled = false
        barChartView.xAxis.labelCount = 5
        barChartView.xAxis.granularity = 1.0
        barChartView.xAxis.labelPosition = .bottom
        
        
        
        // setup realm
        let realm = RLMRealm.default()
        
        realm.beginWriteTransaction()
        
        // clear previous scores that might exist from previous viewing of this VC
        realm.deleteObjects(RealmDemoData.allObjects())
        
        // write some demo-data into the realm.io database
        let score1 = Score(totalScore: 100.0, scoreNr: 0.0, playerName: "Peter")
        realm.add(score1)
        
        let score2 = Score(totalScore: 110.0, scoreNr: 1.0, playerName: "Lisa")
        realm.add(score2)
        
        let score3 = Score(totalScore: 130.0, scoreNr: 2.0, playerName: "Dennis")
        realm.add(score3)
        
        let score4 = Score(totalScore: 70.0, scoreNr: 3.0, playerName: "Luke")
        realm.add(score4)
        
        let score5 = Score(totalScore: 80.0, scoreNr: 4.0, playerName: "Sarah")
        realm.add(score5)
        
        // commit changes to realm db
        _ = try? realm.commitWriteTransaction()
        
        // add data to the chart
        setData()
        
    }
    
    func setData()
    {
        let realm = RLMRealm.default()
        
        results = Score.allObjects(in: realm) as? RLMResults<Score>
        lineChartView.xAxis.valueFormatter = self
        barChartView.xAxis.valueFormatter = self
        
        // Line chart
        let lineDataSet = RealmLineDataSet(results: (results as? RLMResults<RLMObject>), xValueField: "scoreNr", yValueField: "totalScore")
        
        lineDataSet.mode = .linear
        lineDataSet.label = "Result Scores"
        lineDataSet.drawCircleHoleEnabled = false
        lineDataSet.setColor (ChartColorTemplates.colorFromString( "#FF5722"))
        lineDataSet.setCircleColor  ( ChartColorTemplates.colorFromString( "#FF5722"))
        lineDataSet.lineWidth = 1.8
        lineDataSet.circleRadius = 3.6
        
        let lineDataSets: [IChartDataSet] = [lineDataSet]
        let lineData = LineChartData(dataSets: lineDataSets)
        styleData(lineData)
        
        // set data
        lineChartView.data = lineData
        lineChartView.animate(yAxisDuration: 1.4, easingOption: .easeOutQuad)
        
        // Bar chart
        let barDataSet = RealmBarDataSet(results: (results as? RLMResults<RLMObject>), xValueField: "scoreNr", yValueField: "totalScore", label: "")
        barDataSet.colors = [ChartColorTemplates.colorFromString( "#FF5722"), ChartColorTemplates.colorFromString("#03A9F4")]
        barDataSet.label = "Realm BarDataSet"
        let barDataSets: [IChartDataSet] = [barDataSet]
        
        let barData = BarChartData(dataSets: barDataSets)
        styleData(barData)
        barChartView.data = barData
        
        barChartView.fitBars = true
        barChartView.notifyDataSetChanged()
        barChartView.animate(yAxisDuration: 1.4, easingOption: .easeOutQuad)
    }
    
    func optionTapped( sender: NSMenuItem)
    {
        switch (sender.title)
        {
        default:
            super.toggle(sender.title, chartView: lineChartView)
        }
    }
}

// MARK: - ChartViewDelegate
extension RealmWikiExampleChartViewController: ChartViewDelegate
{
    public func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight)
    {
        print("chartValueSelected : x = \(highlight.x)")
    }
    
    public func chartValueNothingSelected(_ chartView: ChartViewBase)
    {
        print("chartValueNothingSelected")
    }
}

extension RealmWikiExampleChartViewController: IAxisValueFormatter
{
    func stringForValue(_ value: Double, axis: AxisBase?) -> String
    {
        return results!.object(at: UInt(value)).playerName
    }
}

